### PR TITLE
feat: store ticket info in dynamodb 

### PR DESCRIPTION
**Pull request by erlend-vimond**
## *What* does this PR do?
- Adds ticket-storer and ticket-storer-pr actions.
  - Ticket-storer is triggered on each push and stores the related tickets in DynamoDB
  - Ticket-storer-pr stores the tickets related to a PR whenever it is opened, changed, closed etc.

## *Why* are you suggesting this change?
This is the start of the compliance workflow.

## *Who* is the change affecting?
Everyone using the GH actions workflow, currently noone.

## *What* is affected by this change?
The vimond-central workflow.
 
## JIRA Tickets
PLAT-564

## Dependencies
None

## Known risks
We are putting (very limited) AWS creds in GH secrets.

## Automated Testing
None

## Manual Testing
Seems to work fine for the testing-terraform-management project: https://github.com/vimond/testing-terraform-management/runs/7648570541?check_suite_focus=true

### CODE REVIEW

File: ticket-storer-pr/src/dynamodb.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+const { DynamoDBClient, PutItemCommand } = require("@aws-sdk/client-dynamodb");
+
+async function storePRTickets(awsConfig, owner, repoName, prID, tickets) {
+    const client = new DynamoDBClient();
+
+    return await client.send(new PutItemCommand({
+        Item: {
+            "OwnerRepo": { S: owner + ":" + repoName },
+            "ID": { S: `PR:${prID}` },
```
Comment (dagvl) [inline ticket-storer-pr/src/dynamodb.js:9]: style: Mixing string concatenation and string templating

File: ticket-storer/index.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    const searchClient = prTicketSearcher.getSearchClient(githubConfig);
+
+    let commitTickets = {};
+    let commitPrs = {};
+    let allTickets = [];
+    let validTickets = {};
+    for (const commit of input.commits) {
+      const tickets = ticketFinder.findAll([commit.message]); // Finding tickets in the commit message
+      commitTickets[commit.id] = tickets;
+      allTickets = allTickets.concat([...tickets]);
```
Comment (dagvl) [inline ticket-storer/index.js:64]: I think idiomatic js would be:
```
allTickets = allTickets.concat(...tickets);
```
example:
```
> a = Array(1,2)
[ 1, 2 ]
> b = new Set()
Set(0) {}
> b.add(3)
Set(1) { 3 }
> b.add(4)
Set(2) { 3, 4 }
> a
[ 1, 2 ]
> a.concat(...b)
[ 1, 2, 3, 4 ]
```

File: ticket-storer/index.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
+const ticketSender = require('./src/dynamodb');
+const core = require('@actions/core');
+const ticketFinder = require('../ticket-collector/src/tickets-finder');
```
Comment (dagvl) [inline ticket-storer/index.js:3]: Common code should be factored out of the action and put on a top level so it is clear for the next person modifying the `tickets-finder` action that some other action also uses this code

File: ticket-storer/index.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+    // Hide secrets
+    core.setSecret(jiraConfig.token);
+    core.setSecret(jiraConfig.username);
+    core.setSecret(githubConfig.token);
+
+
+    if (input.overrideRepo != undefined && input.overrideRepo !== "") {
+      const resp = await ticketSender.storeOverrideRepoName(awsConfig, input.owner, input.repo, input.overrideRepo);
+      console.log("Storing overriden repo name", resp);
```
Comment (dagvl) [inline ticket-storer/index.js:49]: typo

File: ticket-storer/index.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      searchBaseUrl: core.getInput('git-search-base-url')
+    }
+
+    // Hide secrets
+    core.setSecret(jiraConfig.token);
+    core.setSecret(jiraConfig.username);
+    core.setSecret(githubConfig.token);
+
+
+    if (input.overrideRepo != undefined && input.overrideRepo !== "") {
```
Comment (dagvl) [inline ticket-storer/index.js:47]: `!=` -> `!==`

File: ticket-storer/index.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      commitPrs[commit.id] = await prTicketSearcher.searchForCommitPullRequest(searchClient, commit.id);
+    }
+
+    console.log("allTickets", allTickets)
+    let filteredTickets;
+    if (jiraConfig.host !== undefined && jiraConfig.token !== undefined && jiraConfig.username !== undefined) {
+      try {
+        filteredTickets = await jira.checkIfExist(jiraConfig, allTickets);
+        if (filteredTickets == null || typeof filteredTickets[Symbol.iterator] !== 'function') {
+          console.log(filteredTickets);
```
Comment (dagvl) [inline ticket-storer/index.js:77]: I think this is ok here, so we don't overload Jira with concurrent requests, but if you'd want these to run concurrently, you'd store all the Promise objects in an array and use Promise.all(x) on that array.

File: ticket-storer/src/dynamodb.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+// The original repo name is needed to clone the repo when getting the commits between two deployments
+async function storeOverrideRepoName(awsConfig, owner, repoName, overrideRepoName) {
+    const client = new DynamoDBClient();
+
+    return await client.send(new PutItemCommand({
+        Item: {
+            "OwnerRepo": { S: owner + ":" + overrideRepoName },
+            "ID": { S: `reponame` },
```
Comment (dagvl) [inline ticket-storer/src/dynamodb.js:52]: Unneeded templating of static string

File: ticket-storer-pr/index.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      prID: core.getInput("pr-id", { required: true }),
+      inputFile: core.getInput("input-file"),
+      tickets: core.getInput("tickets"),
+    };
+
+    const awsConfig = {
+      tableName: core.getInput("dynamodb-tablename", { required: true }),
+    };
+
+    if (input.overrideRepo != undefined && input.overrideRepo !== "") {
```
Comment (dagvl) [inline ticket-storer-pr/index.js:22]: `!=` -> `!==`

**Review by dagvl — CHANGES_REQUESTED** [review]
Mostly style and minor stuff. Only major thing is that we should refactor out shared code so it is clear that it is shared.

File: ticket-storer-pr/action.yml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+name: 'ticket-collector'
```
[the remark this answers is not in the document]
Comment (erlend-vimond) [reply ticket-storer-pr/action.yml]: Fixed 👍

File: ticket-storer/action.yml [hunk 10]
```diff
@@ -0,0 +1,54 @@
+name: 'ticket-collector'
+description: 'Collects ticket ids'
```
[the remark this answers is not in the document]
Comment (erlend-vimond) [reply ticket-storer/action.yml]: Fixed 👍

File: ticket-storer-pr/index.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      input.repo = input.overrideRepo;
+    }
+
+    console.log("convertig tickets");
```
[the remark this answers is not in the document]
Comment (erlend-vimond) [reply ticket-storer-pr/index.js]: Fixed 👍

File: ticket-storer-pr/src/dynamodb.js (hunk 1 above)
Comment (erlend-vimond) [reply ticket-storer-pr/src/dynamodb.js]: Converted everything to use string templating 👍

File: ticket-storer/index.js (hunk 2 above)
Comment (erlend-vimond) [reply ticket-storer/index.js]: I'm not familiar enough with js to tell, thanks for letting me know! Fixed 👍

File: ticket-storer/index.js (hunk 3 above)
Comment (erlend-vimond) [reply ticket-storer/index.js]: Moved the it to a top level "utils" folder, please review.

File: ticket-storer/index.js (hunk 4 above)
Comment (erlend-vimond) [reply ticket-storer/index.js]: Fixed 👍

File: ticket-storer/index.js (hunk 5 above)
Comment (erlend-vimond) [reply ticket-storer/index.js]: Fixed 👍

File: ticket-storer/index.js (hunk 6 above)
Comment (erlend-vimond) [reply ticket-storer/index.js:77]: Doesn't the JIRA search only return one promise? Did you mean the GitHub PR search?

File: ticket-storer/src/dynamodb.js (hunk 7 above)
Comment (erlend-vimond) [reply ticket-storer/src/dynamodb.js]: Fixed 👍

File: ticket-storer-pr/index.js (hunk 8 above)
Comment (erlend-vimond) [reply ticket-storer-pr/index.js]: Fixed 👍

File: ticket-storer/index.js (hunk 6 above)
Comment (dagvl) [reply ticket-storer/index.js:77]: I think i was mixing up two pieces of code, i was sure there was a loop here :D Just ignore me...

**Review by dagvl — APPROVED** [review]

**Pull request merged**